subprojects/core/src/test/groovy/org/gradle/internal/classpath/CallInterceptingMetaClassTest.groovy

```diff
@@ -0,0 +1,342 @@
+ * limitations under the License.
+ */
+
+package org.gradle.internal.classpath
+
+import org.gradle.internal.classpath.intercept.CallInterceptorResolver
+import org.gradle.internal.classpath.intercept.CallInterceptorsSet
+import spock.lang.Specification
+
+import static org.gradle.internal.classpath.BasicCallInterceptionTestInterceptorsDeclaration.GROOVY_GENERATED_CLASS
+import static org.gradle.internal.classpath.InstrumentedGroovyCallsTracker.CallKind.GET_PROPERTY
+import static org.gradle.internal.classpath.InstrumentedGroovyCallsTracker.CallKind.INVOKE_METHOD
+import static org.gradle.internal.classpath.InstrumentedGroovyCallsTracker.CallKind.SET_PROPERTY
+
+class CallInterceptingMetaClassTest extends Specification {
+
+    CallInterceptorResolver callInterceptors = new CallInterceptorsSet(GroovyCallInterceptorsProvisionTools.getInterceptorsFromClass(GROOVY_GENERATED_CLASS).stream())
+    private MetaClass originalMetaClass = null
+    private static InterceptorTestReceiver instance = null
+
+    // We don't want to interfere with other tests that modify the meta class
+    def interceptorTestReceiverClassLock = new ClassBasedLock(InterceptorTestReceiver)
+
+    def callTracker = new DefaultInstrumentedGroovyCallsTracker()
+
+    def setup() {
+        interceptorTestReceiverClassLock.lock()
+        originalMetaClass = InterceptorTestReceiver.metaClass
+        GroovySystem.metaClassRegistry.setMetaClass(
+            InterceptorTestReceiver,
+            new CallInterceptingMetaClass(GroovySystem.metaClassRegistry, InterceptorTestReceiver, InterceptorTestReceiver.metaClass, callTracker, callInterceptors)
+        )
+        instance = new InterceptorTestReceiver()
+    }
+
+    def cleanup() {
+        GroovySystem.metaClassRegistry.setMetaClass(InterceptorTestReceiver, originalMetaClass)
+        instance = null
+        interceptorTestReceiverClassLock.unlock()
+    }
+
+    def 'intercepts a dynamic call with no argument'() {
+        when:
+        withEntryPoint(INVOKE_METHOD, "test") {
+            instance.invokeMethod("test", [].toArray())
+        }
+
+        then:
+        instance.intercepted == "test()"
+    }
+
+    def 'intercepts a dynamic call with argument'() {
+        when:
+        withEntryPoint(INVOKE_METHOD, "test") {
+            instance.invokeMethod("test", [].toArray())
+        }
+
+        then:
+        instance.intercepted == "test()"
+    }
+
+    def 'intercepts a dynamic call to a non-existent method'() {
+        when:
+        withEntryPoint(INVOKE_METHOD, "nonExistent") {
+            instance.nonExistent("test")
+        }
+
+        then:
+        instance.intercepted == "nonExistent(String)-non-existent"
+
+        when: "calling it outside the entry point scope"
+        instance.nonExistent("test")
+
+        then: "it should throw an exception as usual"
+        thrown(MissingMethodException)
+    }
+
+    def 'intercepts only one call in a succession with a single entry point'() {
```

Comment:
❓ Why?

Reply:
It is the conservative approach: we only intercept calls whose call sites we are confident about. For example, if there is another recursive call inside the original call, and if we did not process the recursive call's call site, then we should not intercept it. For each "normal" call from the build logic classes, we are going to have a new correctly tracked entry point.